src/hypothesis/internal/charstree.py

```diff
@@ -0,0 +1,310 @@
+                    locp, hicp = item[0] + this_base, item[1] + this_base
+                    if locp == codepoint == hicp:
+                        continue
+                    elif not (locp <= codepoint <= hicp):
+                        value_acc.append(item)
+                    elif locp == codepoint:
+                        item = (codepoint + 1 - this_base, item[1])
+                        value_acc.append(item)
+                    elif hicp == codepoint:
+                        item = (item[0], codepoint - 1 - this_base)
+                        value_acc.append(item)
+                    else:
+                        value_acc.append((item[0], codepoint - 1 - this_base))
+                        value_acc.append((codepoint + 1 - this_base, item[1]))
+                filtered_value = value_acc
+            if filtered_value:
+                acc[key] = tuple(filtered_value)
+    return acc
+
+
+def select_values(tree, category, selector):
+    if category not in tree:
+        return 0, []
+    base = 0
+    namespace = tree[category]
+    while True:
+        key = selector(namespace)
+        namespace = namespace[key]
+        base += key
+        if isinstance(namespace, dict):
+            continue
+        return base, namespace
+
+
+def iter_values(namespace, base=0):
+    if isinstance(namespace, dict):
+        for key in namespace:
+            for value in iter_values(namespace[key], base + key):
+                yield value
+    else:
+        for value in namespace:
+            yield (base + value[0], base + value[1])
+
+
+def make_tree():
+    def new_tree():
+        tree = {}
+        for codepoint in range(0, sys.maxunicode + 1):
+            cat = unicodedata.category(hunichr(codepoint))
+            target = tree.setdefault(cat, [])
+            if target and codepoint == target[-1][-1] + 1:
+                target[-1][-1] += 1
+            else:
+                target.append([codepoint, codepoint])
+        return tree
+
+    def fold(tree, factor):
+        for key, values in tree.items():
+            if isinstance(values, dict):
+                fold(values, factor)
+            else:
+                tree[key] = fold_values(values, factor)
+
+    def fold_values(values, factor):
+        group = {}
+        by_factor = lambda i: i[0] & factor
+        for ns, items in itertools.groupby(values, key=by_factor):
+            namespace = group.setdefault(ns, [])
+            namespace.extend([[i[0] - ns, i[1] - ns] for i in items])
+        return group
+
+    tree = new_tree()
+    # Multi folding optimizes file size ~10-15% and improves algorithmic cost
+    # of code point lookup
+    fold(tree, 0xfff000)
+    fold(tree, 0xffff00)
+    fold(tree, 0xfffff0)
+    return tree
+
+
+def cache_file_path():
+    return os.path.join(hypothesis_home_dir(),
+                        os.path.basename(__file__) + '.cache')
+
+
+def dump_tree(tree):  # pragma: no cover
+    dump = marshal.dumps(tree)
```

Comment:
I think I might have been the on who suggested it, but I'm wondering about the validity of using marshal here now. The marshal format is not guaranteed and I think not compatible between Python 2 or 3. This could cause problems for e.g. cross version libraries.

I wonder if using something more portable would make sense. e.g. json is used elsewhere. It would perform worse, but the size isn't large enough that that's likely to be a major problem.